File: sdk/samples/WFSamples/Technologies/Tracking/ConsoleTrackingServiceSample/CS/ConsoleTrackingServiceSample/ConsoleTrackingService.cs
//---------------------------------------------------------------------
//  This file is part of the Windows Workflow Foundation SDK Code Samples.
// 
//  Copyright (C) Microsoft Corporation.  All rights reserved.
// 
//This source code is intended only as a supplement to Microsoft
//Development Tools and/or on-line documentation.  See these other
//materials for detailed information regarding Microsoft code samples.
// 
//THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
//KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
//IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
//PARTICULAR PURPOSE.
//---------------------------------------------------------------------

using System;
using System.IO;
using System.Collections.Generic;
using System.Workflow.Runtime;
using System.Workflow.Runtime.Tracking;
using System.Workflow.ComponentModel;
using System.Workflow.Runtime.Hosting;

namespace Microsoft.Samples.Workflow.ConsoleTrackingServiceSample
{
    public class TrackingChannelSample : TrackingChannel
    {
        private TrackingParameters parameters = null;

        protected TrackingChannelSample()
        {
        }

        public TrackingChannelSample(TrackingParameters parameters)
        {
            this.parameters = parameters;
        }

        // Send() is called by Tracking runtime to send various tracking records
        protected override void Send(TrackingRecord record)
        {
            // filter on record type
            if (record is WorkflowTrackingRecord)
            {
                WriteWorkflowTrackingRecord((WorkflowTrackingRecord)record);
            }
            if (record is ActivityTrackingRecord)
            {
                WriteActivityTrackingRecord((ActivityTrackingRecord)record);
            }
            if (record is UserTrackingRecord)
            {
                WriteUserTrackingRecord((UserTrackingRecord)record);
            }
        }

        // InstanceCompletedOrTerminated() is called by Tracking runtime to indicate that the Workflow instance finished running
        protected override void InstanceCompletedOrTerminated()
        {
            WriteTitle("Workflow Instance Completed or Terminated");
        }

        private static void WriteTitle (string title)
        {
            Console.WriteLine("\n***********************************************************");
            Console.WriteLine("\t" + title);
            Console.WriteLine("***********************************************************");
        }

        private static void WriteWorkflowTrackingRecord(WorkflowTrackingRecord workflowTrackingRecord)
        {
            WriteTitle("Workflow Tracking Record");
            Console.WriteLine("EventDateTime: " + workflowTrackingRecord.EventDateTime.ToString());
            Console.WriteLine("Status: " + workflowTrackingRecord.TrackingWorkflowEvent.ToString());        
        }

        private static void WriteActivityTrackingRecord(ActivityTrackingRecord activityTrackingRecord)
        {
            WriteTitle("Activity Tracking Record");
            Console.WriteLine("EventDateTime: " + activityTrackingRecord.EventDateTime.ToString());
            Console.WriteLine("QualifiedName: " + activityTrackingRecord.QualifiedName.ToString());
            Console.WriteLine("Type: " + activityTrackingRecord.ActivityType);
            Console.WriteLine("Status: " + activityTrackingRecord.ExecutionStatus.ToString());
        }

        private static void WriteUserTrackingRecord(UserTrackingRecord userTrackingRecord)
        {
            WriteTitle("User Activity Record"); 
            Console.WriteLine("EventDateTime: " + userTrackingRecord.EventDateTime.ToString());
            Console.WriteLine("QualifiedName: " + userTrackingRecord.QualifiedName.ToString());
            Console.WriteLine("ActivityType: " + userTrackingRecord.ActivityType.FullName.ToString());
            Console.WriteLine("Args: " + userTrackingRecord.UserData.ToString());
        }
    }

    public class ConsoleTrackingService : TrackingService
    {
        protected override bool TryGetProfile(Type workflowType, out TrackingProfile profile)
        {
            //Depending on the workflowType, service can return different tracking profiles
            //In this sample we're returning the same profile for all running types
            profile = GetProfile();
            return true;
        }

        protected override TrackingProfile GetProfile(Guid workflowInstanceId)
        {
            // Does not support reloading/instance profiles
            throw new NotImplementedException("The method or operation is not implemented.");
        }

        protected override TrackingProfile GetProfile(Type workflowType, Version profileVersionId)
        {
            // Return the version of the tracking profile that runtime requests (profileVersionId)
            return GetProfile();
        }

        protected override bool TryReloadProfile(Type workflowType, Guid workflowInstanceId, out TrackingProfile profile)
        {
            // Returning false to indicate there is no new profiles
            profile = null;
            return false;
        }

        protected override TrackingChannel GetTrackingChannel(TrackingParameters parameters)
        {
            //return a tracking channel to receive runtime events
            return new TrackingChannelSample(parameters);
        }

        // Profile creation
        private static TrackingProfile GetProfile()
        {
            // Create a Tracking Profile
            TrackingProfile profile = new TrackingProfile();
            profile.Version = new Version("3.0.0");

            // Add a TrackPoint to cover all activity status events
            ActivityTrackPoint activityTrackPoint = new ActivityTrackPoint();
            ActivityTrackingLocation activityLocation = new ActivityTrackingLocation(typeof(Activity));
            activityLocation.MatchDerivedTypes = true;
            WorkflowTrackingLocation wLocation = new WorkflowTrackingLocation();

            IEnumerable<ActivityExecutionStatus> statuses = Enum.GetValues(typeof(ActivityExecutionStatus)) as IEnumerable<ActivityExecutionStatus>;
            foreach (ActivityExecutionStatus status in statuses)
            {
                activityLocation.ExecutionStatusEvents.Add(status);
            }

            activityTrackPoint.MatchingLocations.Add(activityLocation);
            profile.ActivityTrackPoints.Add(activityTrackPoint);

            // Add a TrackPoint to cover all workflow status events
            WorkflowTrackPoint workflowTrackPoint = new WorkflowTrackPoint();
            workflowTrackPoint.MatchingLocation = new WorkflowTrackingLocation();
            foreach (TrackingWorkflowEvent workflowEvent in Enum.GetValues(typeof(TrackingWorkflowEvent)))
            {
                workflowTrackPoint.MatchingLocation.Events.Add(workflowEvent);
            }
            profile.WorkflowTrackPoints.Add(workflowTrackPoint);
            
            // Add a TrackPoint to cover all user track points
            UserTrackPoint userTrackPoint = new UserTrackPoint();
            UserTrackingLocation userLocation = new UserTrackingLocation();
            userLocation.ActivityType = typeof(Activity);
            userLocation.MatchDerivedActivityTypes = true;
            userLocation.ArgumentType = typeof(object);
            userLocation.MatchDerivedArgumentTypes = true;
            userTrackPoint.MatchingLocations.Add(userLocation);
            profile.UserTrackPoints.Add(userTrackPoint);

            return profile;
        }
    }
}

File: sdk/samples/WFSamples/Technologies/Tracking/ConsoleTrackingServiceSample/VB/ConsoleTrackingServiceSample/ConsoleTrackingService.vb
'---------------------------------------------------------------------
'  This file is part of the Windows Workflow Foundation SDK Code Samples.
' 
'  Copyright (C) Microsoft Corporation.  All rights reserved.
' 
'This source code is intended only as a supplement to Microsoft
'Development Tools and/or on-line documentation.  See these other
'materials for detailed information regarding Microsoft code samples.
' 
'THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
'KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
'IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
'PARTICULAR PURPOSE.
'---------------------------------------------------------------------

Imports System
Imports System.IO
Imports System.Collections.Generic
Imports System.Workflow.Runtime
Imports System.Workflow.Runtime.Tracking
Imports System.Workflow.ComponentModel
Imports System.Workflow.Runtime.Hosting
Imports System.Globalization
Imports Microsoft.VisualBasic

Public Class TrackingChannelSample
    Inherits TrackingChannel

    Private parametersValue As TrackingParameters = Nothing

    Protected Sub New()
    End Sub

    Public Sub New(ByVal parameters As TrackingParameters)
        Me.parametersValue = parameters
    End Sub

    ' Send() is called by Tracking runtime to send various tracking records
    Protected Overrides Sub Send(ByVal record As TrackingRecord)
        'filter on record type
        If TypeOf record Is WorkflowTrackingRecord Then
            WriteWorkflowTrackingRecord(CType(record, WorkflowTrackingRecord))
        End If
        If TypeOf record Is ActivityTrackingRecord Then
            WriteActivityTrackingRecord(CType(record, ActivityTrackingRecord))
        End If
        If TypeOf record Is UserTrackingRecord Then
            WriteUserTrackingRecord(CType(record, UserTrackingRecord))
        End If
    End Sub

    ' InstanceCompletedOrTerminated() is called by Tracking runtime to indicate that the Workflow instance finished running
    Protected Overrides Sub InstanceCompletedOrTerminated()
        WriteTitle("Workflow Instance Completed or Terminated")
    End Sub

    Private Shared Sub WriteTitle(ByVal title As String)
        Console.WriteLine(vbCrLf + "***********************************************************")
        Console.WriteLine(vbTab + title)
        Console.WriteLine("***********************************************************")
    End Sub

    Private Shared Sub WriteWorkflowTrackingRecord(ByVal workflowTrackingRecord As WorkflowTrackingRecord)
        WriteTitle("Workflow Tracking Record")
        Console.WriteLine("EventDateTime: " + workflowTrackingRecord.EventDateTime.ToString(CultureInfo.CurrentCulture))
        Console.WriteLine("Status: " + workflowTrackingRecord.TrackingWorkflowEvent.ToString())
    End Sub

    Private Shared Sub WriteActivityTrackingRecord(ByVal activityTrackingRecord As ActivityTrackingRecord)
        WriteTitle("Activity Tracking Record")
        Console.WriteLine("EventDateTime: " + activityTrackingRecord.EventDateTime.ToString(CultureInfo.CurrentCulture))
        Console.WriteLine("QualifiedName: " + activityTrackingRecord.QualifiedName.ToString())
        Console.WriteLine("Type: " + activityTrackingRecord.ActivityType.ToString())
        Console.WriteLine("Status: " + activityTrackingRecord.ExecutionStatus.ToString())
    End Sub

    Private Shared Sub WriteUserTrackingRecord(ByVal userTrackingRecord As UserTrackingRecord)
        WriteTitle("User Activity Record")
        Console.WriteLine("EventDateTime: " + userTrackingRecord.EventDateTime.ToString(CultureInfo.CurrentCulture))
        Console.WriteLine("QualifiedName: " + userTrackingRecord.QualifiedName.ToString())
        Console.WriteLine("ActivityType: " + userTrackingRecord.ActivityType.FullName.ToString())
        Console.WriteLine("Args: " + userTrackingRecord.UserData.ToString())
    End Sub
End Class

Public Class ConsoleTrackingService
    Inherits TrackingService

    Protected Overrides Function TryGetProfile(ByVal workflowType As Type, ByRef profile As TrackingProfile) As Boolean
        'Depending on the workflowType, service can return different tracking profiles
        'In this sample we're returning the same profile for all running types
        profile = GetProfile()
        Return True
    End Function

    Protected Overrides Function GetProfile(ByVal workflowInstanceId As Guid) As TrackingProfile
        ' Does not support reloading/instance profiles
        Throw New NotImplementedException("The method or operation is not implemented.")
    End Function

    Protected Overrides Function GetProfile(ByVal workflowType As Type, ByVal profileVersionId As Version) As TrackingProfile
        ' Return the version of the tracking profile that runtime requests (profileVersionId)
        Return GetProfile()
    End Function

    Protected Overrides Function TryReloadProfile(ByVal workflowType As Type, ByVal workflowInstanceId As Guid, ByRef profile As TrackingProfile) As Boolean
        ' Returning false to indicate there is no new profiles
        profile = Nothing
        Return False
    End Function

    Protected Overrides Function GetTrackingChannel(ByVal parameters As TrackingParameters) As TrackingChannel
        ' return a tracking channel to receive runtime events
        Return New TrackingChannelSample(parameters)
    End Function

    ' Profile creation
    Private Overloads Shared Function GetProfile() As TrackingProfile
        ' Create a Tracking Profile
        Dim profile As TrackingProfile = New TrackingProfile()
        profile.Version = New Version("3.0.0")

        ' Add a TrackPoint to cover all activity status events
        Dim activityTrackPoint As ActivityTrackPoint = New ActivityTrackPoint()
        Dim activityLocation As ActivityTrackingLocation = New ActivityTrackingLocation(GetType(Activity))
        activityLocation.MatchDerivedTypes = True
        Dim statuses As IEnumerable(Of ActivityExecutionStatus) = CType(System.Enum.GetValues(GetType(ActivityExecutionStatus)), IEnumerable(Of ActivityExecutionStatus))
        For Each status As ActivityExecutionStatus In statuses
            activityLocation.ExecutionStatusEvents.Add(status)
        Next
        activityTrackPoint.MatchingLocations.Add(activityLocation)
        profile.ActivityTrackPoints.Add(activityTrackPoint)

        ' Add a TrackPoint to cover all workflow status events
        Dim workflowTrackPoint As WorkflowTrackPoint = New WorkflowTrackPoint()
        workflowTrackPoint.MatchingLocation = New WorkflowTrackingLocation()
        For Each workflowEvent As TrackingWorkflowEvent In System.Enum.GetValues(GetType(TrackingWorkflowEvent))
            workflowTrackPoint.MatchingLocation.Events.Add(workflowEvent)
        Next
        profile.WorkflowTrackPoints.Add(workflowTrackPoint)

        ' Add a TrackPoint to cover all user track points
        Dim userTrackPoint As UserTrackPoint = New UserTrackPoint()
        Dim userLocation As UserTrackingLocation = New UserTrackingLocation()
        userLocation.ActivityType = GetType(Activity)
        userLocation.MatchDerivedActivityTypes = True
        userLocation.ArgumentType = GetType(Object)
        userLocation.MatchDerivedArgumentTypes = True
        userTrackPoint.MatchingLocations.Add(userLocation)
        profile.UserTrackPoints.Add(userTrackPoint)

        Return profile
    End Function
End Class